---
- 07/23

In [412]:
import pandas as pd
import numpy as np

#워닝 메시지
import warnings
warnings.filterwarnings('ignore')

In [413]:
df_외감=pd.read_csv('./Dataset/외감raw3.csv',encoding='euc-kr')

In [414]:
# df_외감.drop(df_외감[df_외감['회사명'].isin(['SK리츠운용주식회사', '굿모닝대신리츠주식회사', '(주)동방리츠', '대영리츠건설(주)', '드림리츠(주)',
#        '(주)리츠라인', '(주)리츠산업', '(주)리츠홀딩스', '(주)블리츠웨이', '메리츠건설(주)',
#        '(주)서울리츠임대주택제2호위탁관리부동산투자회사', '신한리츠운용주식회사',
#        '엔에이치농협리츠운용주식회사', '영신리츠팜주식회사', '(주)제이알리츠', '월드리츠개발주식회사',
#        '주식회사경기리츠공공임대제1호위탁관리부동산투자회사', '주식회사경성리츠', '주식회사글리츠리츠',
#        '주식회사대신케이리츠물류1호위탁관리부동산투자회사', '주식회사리츠파크', '주식회사무등리츠', '주식회사비앤에스리츠',
#        '주식회사서울리츠임대주택제1호위탁관리부동산투자회사', '주식회사서울리츠임대주택제3호위탁관리부동산투자회사',
#        '주식회사서울리츠임대주택제4호위탁관리부동산투자회사', '주식회사아이에이치큐리츠', '주식회사와이에스리츠',
#        '주식회사제로에너지임대리츠2호위탁관리부동산투자회사', '주식회사조일리츠', 
#        '주식회사토지지원리츠제1호위탁관리부동산투자회사', '주식회사토지지원리츠제2호위탁관리부동산투자회사', '주식회사평안리츠',
#        '주식회사플리츠', '주식회사하우리츠', '주식회사힘내라리츠', '(주)청리츠앤파트너스',
#        '(주)케이리츠투자운용', '케이리츠주식회사', '켄달스퀘어리츠운용주식회사', '한국리츠주식회사'])].index,axis=0,inplace=True)

list=['SK리츠운용주식회사', '굿모닝대신리츠주식회사', '(주)동방리츠', '대영리츠건설(주)', '드림리츠(주)',
       '(주)리츠라인', '(주)리츠산업', '(주)리츠홀딩스', '(주)블리츠웨이', '메리츠건설(주)',
       '(주)서울리츠임대주택제2호위탁관리부동산투자회사', '신한리츠운용주식회사',
       '엔에이치농협리츠운용주식회사', '영신리츠팜주식회사', '(주)제이알리츠', '월드리츠개발주식회사',
       '주식회사경기리츠공공임대제1호위탁관리부동산투자회사', '주식회사경성리츠', '주식회사글리츠리츠',
       '주식회사대신케이리츠물류1호위탁관리부동산투자회사', '주식회사리츠파크', '주식회사무등리츠', '주식회사비앤에스리츠',
       '주식회사서울리츠임대주택제1호위탁관리부동산투자회사', '주식회사서울리츠임대주택제3호위탁관리부동산투자회사',
       '주식회사서울리츠임대주택제4호위탁관리부동산투자회사', '주식회사아이에이치큐리츠', '주식회사와이에스리츠',
       '주식회사제로에너지임대리츠2호위탁관리부동산투자회사', '주식회사조일리츠', 
       '주식회사토지지원리츠제1호위탁관리부동산투자회사', '주식회사토지지원리츠제2호위탁관리부동산투자회사', '주식회사평안리츠',
       '주식회사플리츠', '주식회사하우리츠', '주식회사힘내라리츠', '(주)청리츠앤파트너스',
       '(주)케이리츠투자운용', '케이리츠주식회사', '켄달스퀘어리츠운용주식회사', '한국리츠주식회사']
for i in list:
       df_외감 = df_외감[~df_외감['회사명'].str.contains(i)]


In [415]:
# df_외감.drop(df_외감[df_외감['회사명'].isin(['HSBC펀드서비스(주)', '주식회사어니스트펀드','주식회사인천투자펀드2', '주식회사피플펀드컴퍼니'])].index,axis=0,inplace=True)

list =['HSBC펀드서비스(주)', '주식회사어니스트펀드','주식회사인천투자펀드2', '주식회사피플펀드컴퍼니']

for i in list:
    df_외감 = df_외감[~df_외감['회사명'].str.contains(i)]
    

In [416]:
# df_외감.drop(df_외감[df_외감['회사명'].isin(['(주)건설투자', '(주)국제선박투자운용','(유)비티투자','스마트투자파트너스주식회사', '(주)에스와이투자자문', '신동해투자일호피에프브이(주)',
#        '아이비투자검단(주)', '아이비투자을지로주식회사', '에이스투자(주)', '에이치티투자금융주식회사',
#        '엠제이투자시스템주식회사', '유한회사한일자산관리앤투자', '인터씨티개발투자(주)', '일조투자디앤씨(주)',
#        '주식회사건인투자', '주식회사드림부동산투자연구소', '주식회사세원투자건업', '주식회사세원투자산업',
#        '주식회사아프로투자', '주식회사에이디투자', '주식회사에이치엠지기술투자',
#        '주식회사이상투자그룹', '주식회사제이에스투자원', '주식회사케이티투자운용',
#        '주식회사테라투자파트너스', '주식회사한국민간인프라투자', '주식회사한우즈벡실린더투자', '주식회사화신투자',
#        '(주)투자가이드', '케이디기술투자(주)', '케이엔씨개발투자컨설팅', '코레이트투자운용주식회사', '(주)한진정석투자',
#        '(주)현대부동산지분투자', '포항프라이머스프로젝트투자금융주식회사', '한원광장프로젝트투자금융주식회사'])].index,axis=0,inplace=True)

list=['(주)건설투자', '(주)국제선박투자운용','(유)비티투자','스마트투자파트너스주식회사', '(주)에스와이투자자문', '신동해투자일호피에프브이(주)',
       '아이비투자검단(주)', '아이비투자을지로주식회사', '에이스투자(주)', '에이치티투자금융주식회사',
       '엠제이투자시스템주식회사', '유한회사한일자산관리앤투자', '인터씨티개발투자(주)', '일조투자디앤씨(주)',
       '주식회사건인투자', '주식회사드림부동산투자연구소', '주식회사세원투자건업', '주식회사세원투자산업',
       '주식회사아프로투자', '주식회사에이디투자', '주식회사에이치엠지기술투자',
       '주식회사이상투자그룹', '주식회사제이에스투자원', '주식회사케이티투자운용',
       '주식회사테라투자파트너스', '주식회사한국민간인프라투자', '주식회사한우즈벡실린더투자', '주식회사화신투자',
       '(주)투자가이드', '케이디기술투자(주)', '케이엔씨개발투자컨설팅', '코레이트투자운용주식회사', '(주)한진정석투자',
       '(주)현대부동산지분투자', '포항프라이머스프로젝트투자금융주식회사', '한원광장프로젝트투자금융주식회사']
for i in list:
    df_외감 = df_외감[~df_외감['회사명'].str.contains(i)]
    

# df_Outside.drop(df_Outside[df_Outside['Company Name'].isin(값_리스트)].index, axis=0, inplace=True)

In [417]:
# df_외감.drop(df_외감[df_외감['회사명'].isin(df_외감['회사명'].str.contains('부동산투자회사'))].index, axis=0, inplace=True)
# df_외감.drop(df_외감[df_외감['회사명'].isin(df_외감['회사명'].str.contains('금융투자주식회사'))].index, axis=0, inplace=True)
# df_외감.drop(df_외감[df_외감['회사명'].isin(df_외감['회사명'].str.contains('선박투자회사'))].index, axis=0, inplace=True)
# df_외감.drop(df_외감[df_외감['회사명'].isin(df_외감['회사명'].str.contains('투자회사'))].index, axis=0, inplace=True)
# df_외감.drop(df_외감[df_외감['회사명'].isin(df_외감['회사명'].str.contains('투자주식회사'))].index, axis=0, inplace=True)
# df_외감.drop(df_외감[df_외감['회사명'].isin(df_외감['회사명'].str.contains('투자개발'))].index, axis=0, inplace=True)
# df_외감.drop(df_외감[df_외감['회사명'].isin(df_외감['회사명'].str.contains('부동산투지회사'))].index, axis=0, inplace=True)
# df_외감.drop(df_외감[df_외감['회사명'].isin(df_외감['회사명'].str.contains('금융투자'))].index, axis=0, inplace=True)

df_외감 = df_외감[~df_외감['회사명'].str.contains('부동산투자회사')]
df_외감 = df_외감[~df_외감['회사명'].str.contains('금융투자주식회사')]
df_외감 = df_외감[~df_외감['회사명'].str.contains('선박투자회사')]
df_외감 = df_외감[~df_외감['회사명'].str.contains('투자회사')]
df_외감 = df_외감[~df_외감['회사명'].str.contains('투자주식회사')]
df_외감 = df_외감[~df_외감['회사명'].str.contains('투자개발')]
df_외감 = df_외감[~df_외감['회사명'].str.contains('부동산투지회사')]
df_외감 = df_외감[~df_외감['회사명'].str.contains('금융투자')]

In [418]:
df_외감.isna().sum()

회사명                                       0
거래소코드                                     0
회계년도                                      0
통계청 한국표준산업분류 코드 10차(대분류)                  0
부채비율                                     80
당좌비율                                     80
유동비율                                     80
이자보상배율(이자비용)                             80
영업활동으로 인한 현금흐름(간접법)(*)(IFRS연결)(천원)       69
이자비용(IFRS연결)(천원)                         69
차입금의존도                                   80
자기자본구성비율                                 80
당기순이익(손실)(IFRS연결)(천원)                    69
감가상각비(백만원)                             2000
금융비용(백만원)                               300
단기차입금(*)(IFRS연결)(천원)                     69
유동성장기부채(*)(IFRS연결)(천원)                   69
장기차입금(*)(IFRS연결)(천원)                     69
법인세비용차감전(계속사업)손익(백만원)                   103
매출액영업이익률                                 87
자기자본순이익률                                 87
총자본순이익률                                  87
매출액(수익)(*)(IFRS연결)(천원)          

In [419]:
df_외감['감사의견코드']=df_외감['감사의견코드'].map({
    'AG': 1,
    'DS':1,
    'DI':1,
    'DU':1,
    'QA':0,
    'QC':0,
    'QG':0,
    'QQ':0,
    'QS':1,
    'QU':0,
    'UQ':0,
    'ZZ':1
})

## 자본잠식

In [420]:
df_외감['자본잠식률']=(df_외감['자본금(*)(IFRS연결)(천원)'] - df_외감['자본(*)(IFRS연결)(천원)'])/df_외감['자본금(*)(IFRS연결)(천원)']

In [421]:
mask = (df_외감['거래소코드'] == df_외감['거래소코드'].shift(-1)) & (df_외감['자본잠식률'] >= 50)

In [422]:
df_외감['감사의견코드'].value_counts()

감사의견코드
0.0    239093
1.0     20893
Name: count, dtype: int64

In [423]:
df_외감.loc[mask, '감사의견코드'] = 1.0

In [424]:
mask.value_counts()

False    289149
True       4809
Name: count, dtype: int64

In [425]:
df_외감['감사의견코드'].value_counts()

감사의견코드
0.0    236292
1.0     24748
Name: count, dtype: int64

### 자본잠식 > 50인 3951개가 0에서 1로 레이블링 변경

## 회생절차 기업 확인하기!

In [426]:
# 회생절차 개시신청
df_외감.loc[132330,'감사의견코드']=1.0
df_외감.loc[12111,'감사의견코드']=1.0
df_외감.loc[16614,'감사의견코드']=1.0
df_외감.loc[77037,'감사의견코드']=1.0
df_외감.loc[191585,'감사의견코드']=1.0
df_외감.loc[77037,'감사의견코드']=1.0
df_외감.loc[211114,'감사의견코드']=1.0
df_외감.loc[228410,'감사의견코드']=1.0
df_외감.loc[256817,'감사의견코드']=1.0
df_외감.loc[280187,'감사의견코드']=1.0
df_외감.loc[288076,'감사의견코드']=1.0
# 해산사유 발생
# 77485 (주)세실 2017년 파산  ->2016년 레이블링 1.0으로 수정 필요


In [427]:
df_외감['감사의견코드'].value_counts()

감사의견코드
0.0    236282
1.0     24758
Name: count, dtype: int64

### 오름차순으로 정렬 회사명, 거래소 코드로

In [428]:
df_외감.sort_values(by=['회사명','거래소코드'], ascending=True,inplace=True)

### 회사명이 한개인 기업들 삭제

In [429]:
df_외감=df_외감[df_외감['거래소코드'].map(df_외감['거래소코드'].value_counts()) != 1]


In [430]:
df_외감['감사의견코드'].value_counts()

감사의견코드
0.0    234559
1.0     24140
Name: count, dtype: int64

In [431]:
df_외감['Year']= df_외감['회계년도'].replace('/','').astype(str).str[:4].astype(int)

In [432]:
pd.reset_option('display.max_rows', None)
df_외감[(df_외감['거래소코드'] == df_외감['거래소코드'].shift(-1)) & (df_외감['Year'].diff(-1) != -1) &(df_외감['감사의견코드'] ==1) &(df_외감['Year'] !=2012)]

회사명  거래소코드     회계년도  통계청 한국표준산업분류 코드 10차(대분류)     부채비율   
48342       (유)딘텍스코리아  78313  2016/12                        14   478.02  \
1014    (주)강남터미널지하쇼핑몰  65136  2014/12                        68     0.00   
1965       (주)경동스마트산업  80745  2017/12                        41   124.08   
2537        (주)경흥아이앤씨  89843  2019/12                        23  1285.04   
3175       (주)골든블루마리나  83246  2017/12                        91     0.00   
...               ...    ...      ...                       ...      ...   
292080      호진산업기연(주)  69425  2014/12                        27    32.72   
294089       화성고속주식회사  76910  2016/12                        49   170.53   
294356       화신자원주식회사  73705  2015/12                        38   197.66   
294763        환희개발(주)  70581  2014/12                        23    35.58   
294980       효성금속주식회사  86930  2018/12                        24   173.51   

          당좌비율    유동비율  이자보상배율(이자비용)  영업활동으로 인한 현금흐름(간접법)(*)(IFRS연결)(천원)   
48342    34.36  100.88  1.420000e+00                            879590.0  \
1014    285.49  285.49 -1.320800e+02                          -1649243.0   
1965    183.47  183.47  3.072200e+02                             99547.0   
2537      3.83    3.83 -6.870000e+00                          -1753980.0   
3175    148.73  148.73  1.000000e+09                            252423.0   
...        ...     ...           ...                                 ...   
292080  469.32  558.90  8.400000e+00                            579575.0   
294089   66.84   79.98  7.620000e+00                          -6076371.0   
294356  199.06  305.28 -1.750000e+00                           1636479.0   
294763  182.88  211.72  1.084000e+01                           2919006.0   
294980   89.51   96.63  6.990000e+00                           1921453.0   

        이자비용(IFRS연결)(천원)  ...  연구개발비(IFRS연결)(천원)  감사의견코드   
48342           506952.0  ...       7.480000e+10     1.0  \
1014              1383.0  ...      -1.830000e+10     1.0   
1965              3538.0  ...       1.087000e+11     1.0   
2537            339910.0  ...      -2.336000e+11     1.0   
3175                 0.0  ...      -9.180000e+10     1.0   
...                  ...  ...                ...     ...   
292080          104966.0  ...       9.510000e+10     1.0   
294089           40517.0  ...       3.090000e+10     1.0   
294356          262216.0  ...      -4.600000e+10     1.0   
294763           96676.0  ...       1.048000e+11     1.0   
294980          233720.0  ...       1.633000e+11     1.0   

          통계청 한국표준산업분류 10차(대분류)  자본금(*)(IFRS연결)(천원)   
48342                       제조업            100000.0  \
1014                       부동산업             13703.0   
1965                        건설업           1210000.0   
2537                        제조업           3000000.0   
3175        예술, 스포츠 및 여가관련 서비스업            650000.0   
...                         ...                 ...   
292080                      제조업            450000.0   
294089                 운수 및 창고업            625000.0   
294356  수도, 하수 및 폐기물 처리, 원료 재생업            900000.0   
294763                      제조업           1300000.0   
294980                      제조업            500000.0   

        * (정상)영업손익(보고서기재)(IFRS연결)(천원)          회사명.1   회계년도.1  연구개발비(천원)   
48342                        719125.0      (유)딘텍스코리아  2016/12    28940.0  \
1014                        -182670.0  (주)강남터미널지하쇼핑몰  2014/12        0.0   
1965                        1086946.0     (주)경동스마트산업  2017/12        0.0   
2537                       -2335636.0      (주)경흥아이앤씨  2019/12        0.0   
3175                        -917974.0     (주)골든블루마리나  2017/12        0.0   
...                               ...            ...      ...        ...   
292080                       882081.0      호진산업기연(주)  2014/12    69377.0   
294089                       308849.0       화성고속주식회사  2016/12        0.0   
294356                      -460160.0       화신자원주식회사  2015/12        0.0   
294763                      1048413.0        환희개발(

In [433]:
df_외감[~(df_외감['회사명'] == df_외감['회사명'].shift(-1)) | (abs(df_외감['Year']-df_외감['Year'].shift(-1)) == 1)]['감사의견코드'].value_counts()

감사의견코드
0.0    231716
1.0     23583
Name: count, dtype: int64

In [434]:
# df_외감.to_csv('t시점_외감.csv',index=False,encoding='euc-kr')

----
- 07/24

- df_외감 : (291589,55)
- 감사의견코드 결측치 : 32890
- 감사의견코드 value_counts
    - true  : 234559
    - false : 24140

---

* 기업의 target : 0 이면 --> 그 기업 데이터 모두 유지
* 기업의 target : 1 이면 --> 그 기업 1 이전 데이터 유지 & 1 이후 데이터 삭제
* 삭제하고 기업의 개수가 1개인 기업은 drop!

In [435]:
import pandas as pd

# 거래소코드로 그룹화하여 처리하는 함수
def process_group(group):
    # 빈 그룹인 경우 그대로 반환
    if group.empty:
        return group
    
    # 거래소코드 그룹별로 처리
    if 1 in group['감사의견코드'].values:
        # 1이 처음으로 나오는 인덱스 찾기
        first_index = group.index[group['감사의견코드'] == 1][0]
        
        # 처음으로 1이 나오는 행의 아래 행들 삭제
        group = group.loc[:first_index]
    else:
        # 1이 없는 경우 그룹 그대로 유지
        pass
    
    return group



In [436]:
# '거래소코드'로 그룹화하여 함수를 적용하고 다시 합치기
processed_df = df_외감.groupby('거래소코드').apply(process_group)

processed_df

회사명   거래소코드     회계년도  통계청 한국표준산업분류 코드 10차(대분류)    부채비율   
거래소코드                                                                         
10000  269007   (주)포스코건설   10000  2012/12                        41  173.49  \
       269008   (주)포스코건설   10000  2013/12                        41  130.89   
       269009   (주)포스코건설   10000  2014/12                        41  120.92   
       269010   (주)포스코건설   10000  2015/12                        41  100.59   
       269011   (주)포스코건설   10000  2016/12                        41  145.02   
...                  ...     ...      ...                       ...     ...   
107274 215211  주식회사이안이앤디  107274  2021/12                        68    0.00   
107276 216611    주식회사일그램  107276  2021/12                        71   99.28   
       216612    주식회사일그램  107276  2022/12                        71  145.67   
107278 32679   (주)두산종합건재  107278  2021/12                        46  273.78   
       32680   (주)두산종합건재  107278  2022/12                        46  909.68   

                  당좌비율     유동비율  이자보상배율(이자비용)   
거래소코드                                           
10000  269007   124.35   149.99         14.79  \
       269008   120.59   146.23         13.84   
       269009   125.34   153.62          6.33   
       269010   145.66   164.20          7.33   
       269011   109.23   133.86         -8.08   
...                ...      ...           ...   
107274 215211   118.11   497.44         -8.47   
107276 216611  1838.90  1838.90         25.74   
       216612   448.46   448.46         13.22   
107278 32679    376.36   394.26          1.13   
       32680    284.24   291.35         -6.44   

               영업활동으로 인한 현금흐름(간접법)(*)(IFRS연결)(천원)  이자비용(IFRS연결)(천원)  ...   
거래소코드                                                                ...   
10000  269007                         123580817.0        45854000.0  ...  \
       269008                         383458439.0        44520000.0  ...   
       269009                         -81029724.0        55648000.0  ...   
       269010                         645297841.0        61387000.0  ...   
       269011                        -616838633.0        45847000.0  ...   
...                                           ...               ...  ...   
107274 215211                         -17776349.0          549886.0  ...   
107276 216611                           3232717.0          108993.0  ...   
       216612                           3770208.0          315760.0  ...   
107278 32679                           -2107626.0          241431.0  ...   
       32680                             154239.0          328282.0  ...   

               연구개발비(IFRS연결)(천원)  감사의견코드  통계청 한국표준산업분류 10차(대분류)   
거래소코드                                                             
10000  269007       0.000000e+00     0.0                    건설업  \
       269008       0.000000e+00     0.0                    건설업   
       269009       0.000000e+00     0.0                    건설업   
       269010       0.000000e+00     0.0                    건설업   
       269011       0.000000e+00     0.0                    건설업   
...                          ...     ...                    ...   
107274 215211      -4.656000e+11     1.0                   부동산업   
107276 216611       2.805000e+11     NaN       전문, 과학 및 기술 서비스업   
       216612       4.173000e+11     0.0       전문, 과학 및 기술 서비스업   
107278 32679        2.730000e+10     NaN               도매 및 소매업   
       32680       -2.113000e+11     1.0               도매 및 소매업   

               자본금(*)(IFRS연결)(천원)  * (정상)영업손익(보고서기재)(IFRS연결)(천원)      회사명.1   
거래소코드                                                                         
10000  269007         183615000.0                    325556346.0   (주)포스코건설  \
       269008         183615000.0                    448381418.0   (주)포스코건설   
       269009         183615000.0                    322993697.0   (주)포스코건설   
       269010         209033470.0                    138905344.0   (주)포스코건설   
       269011       

In [437]:
processed_df['감사의견코드'].value_counts()

감사의견코드
0.0    199447
1.0     16452
Name: count, dtype: int64

In [438]:
processed_df=processed_df[~(((processed_df['거래소코드']) ==1) & (processed_df['거래소코드'].value_counts()) == 1)]

In [439]:
processed_df

회사명   거래소코드     회계년도  통계청 한국표준산업분류 코드 10차(대분류)    부채비율   
거래소코드                                                                         
10000  269007   (주)포스코건설   10000  2012/12                        41  173.49  \
       269008   (주)포스코건설   10000  2013/12                        41  130.89   
       269009   (주)포스코건설   10000  2014/12                        41  120.92   
       269010   (주)포스코건설   10000  2015/12                        41  100.59   
       269011   (주)포스코건설   10000  2016/12                        41  145.02   
...                  ...     ...      ...                       ...     ...   
107274 215211  주식회사이안이앤디  107274  2021/12                        68    0.00   
107276 216611    주식회사일그램  107276  2021/12                        71   99.28   
       216612    주식회사일그램  107276  2022/12                        71  145.67   
107278 32679   (주)두산종합건재  107278  2021/12                        46  273.78   
       32680   (주)두산종합건재  107278  2022/12                        46  909.68   

                  당좌비율     유동비율  이자보상배율(이자비용)   
거래소코드                                           
10000  269007   124.35   149.99         14.79  \
       269008   120.59   146.23         13.84   
       269009   125.34   153.62          6.33   
       269010   145.66   164.20          7.33   
       269011   109.23   133.86         -8.08   
...                ...      ...           ...   
107274 215211   118.11   497.44         -8.47   
107276 216611  1838.90  1838.90         25.74   
       216612   448.46   448.46         13.22   
107278 32679    376.36   394.26          1.13   
       32680    284.24   291.35         -6.44   

               영업활동으로 인한 현금흐름(간접법)(*)(IFRS연결)(천원)  이자비용(IFRS연결)(천원)  ...   
거래소코드                                                                ...   
10000  269007                         123580817.0        45854000.0  ...  \
       269008                         383458439.0        44520000.0  ...   
       269009                         -81029724.0        55648000.0  ...   
       269010                         645297841.0        61387000.0  ...   
       269011                        -616838633.0        45847000.0  ...   
...                                           ...               ...  ...   
107274 215211                         -17776349.0          549886.0  ...   
107276 216611                           3232717.0          108993.0  ...   
       216612                           3770208.0          315760.0  ...   
107278 32679                           -2107626.0          241431.0  ...   
       32680                             154239.0          328282.0  ...   

               연구개발비(IFRS연결)(천원)  감사의견코드  통계청 한국표준산업분류 10차(대분류)   
거래소코드                                                             
10000  269007       0.000000e+00     0.0                    건설업  \
       269008       0.000000e+00     0.0                    건설업   
       269009       0.000000e+00     0.0                    건설업   
       269010       0.000000e+00     0.0                    건설업   
       269011       0.000000e+00     0.0                    건설업   
...                          ...     ...                    ...   
107274 215211      -4.656000e+11     1.0                   부동산업   
107276 216611       2.805000e+11     NaN       전문, 과학 및 기술 서비스업   
       216612       4.173000e+11     0.0       전문, 과학 및 기술 서비스업   
107278 32679        2.730000e+10     NaN               도매 및 소매업   
       32680       -2.113000e+11     1.0               도매 및 소매업   

               자본금(*)(IFRS연결)(천원)  * (정상)영업손익(보고서기재)(IFRS연결)(천원)      회사명.1   
거래소코드                                                                         
10000  269007         183615000.0                    325556346.0   (주)포스코건설  \
       269008         183615000.0                    448381418.0   (주)포스코건설   
       269009         183615000.0                    322993697.0   (주)포스코건설   
       269010         209033470.0                    138905344.0   (주)포스코건설   
       269011       

In [440]:
processed_df['감사의견코드'].value_counts()

감사의견코드
0.0    199447
1.0     16452
Name: count, dtype: int64

In [441]:
247790-244823

2967

In [442]:
processed_df['감사의견코드'].isna().sum()

31891

In [443]:
mask = processed_df['거래소코드'] == processed_df['거래소코드'].shift(-1)
processed_df['t-1감사의견코드'] = np.where(mask,processed_df['감사의견코드'].shift(-1), np.nan)

In [444]:
processed_df['t-1감사의견코드'].value_counts()

t-1감사의견코드
0.0    182115
1.0     13485
Name: count, dtype: int64

In [445]:
processed_df.dropna(subset='t-1감사의견코드',axis=0,inplace=True)

In [446]:
processed_df.drop(columns='감사의견코드',axis=1,inplace=True)

In [447]:
processed_df.isna().sum()

회사명                                      0
거래소코드                                    0
회계년도                                     0
통계청 한국표준산업분류 코드 10차(대분류)                 0
부채비율                                    60
당좌비율                                    60
유동비율                                    60
이자보상배율(이자비용)                            60
영업활동으로 인한 현금흐름(간접법)(*)(IFRS연결)(천원)      58
이자비용(IFRS연결)(천원)                        58
차입금의존도                                  60
자기자본구성비율                                60
당기순이익(손실)(IFRS연결)(천원)                   58
감가상각비(백만원)                            1385
금융비용(백만원)                              204
단기차입금(*)(IFRS연결)(천원)                    58
유동성장기부채(*)(IFRS연결)(천원)                  58
장기차입금(*)(IFRS연결)(천원)                    58
법인세비용차감전(계속사업)손익(백만원)                   75
매출액영업이익률                                60
자기자본순이익률                                60
총자본순이익률                                 60
매출액(수익)(*)(IFRS연결)(천원)                  58
총자본회전률     

In [448]:
processed_df.drop(['회사명.1','회계년도.1'],axis=1,inplace=True)

In [449]:
processed_df.drop('연구개발비(IFRS연결)(천원)',axis=1,inplace=True)

In [450]:
processed_df.columns=['회사명','거래소코드', '회계년도', '산업분류코드', '부채비율', '당좌비율', '유동비율',
       '이자보상배율', 'OCF(천원)','이자비용(천원)', '차입금의존도', '자기자본구성비율', '당기순이익(천원)',
       '감가상각비(백만원)', '금융비용(백만원)', '단기차입금(천원)','유동성장기부채(천원)', '장기차입금(천원)',
       '법인세비용차감전손익(백만원)', '매출액영업이익률', '자기자본순이익률', '총자본순이익률',
       '매출액(천원)', '총자본회전률', '자기자본회전률', '운전자본회전률', '순운전자본회전률',
       '재고자산회전률', '당좌자산회전률', '유동자산회전률', '매출액증가율', '총자본증가율', '자기자본증가율',
       '순이익증가율', '유형자산증가율', '유동자산증가율', '재고자산증가율', '영업이익증가율', '총자본투자효율',
       '부가가치율', '노동소득분배율', '자본분배율', '이윤분배율', 'ICF(천원)','FCF(천원)', '자산총계(천원)',
       '부채총계(천원)', '자본총계(천원)', 
       '산업분류', '자본금(천원)','영업손익(천원)','연구개발비(천원)','자본잠식률', 'Year', 't-1감사의견코드']

In [451]:
# processed_df.to_csv('t-1_raw_외감.csv',index=False,encoding='euc-kr')

---
- t-2 시점

In [452]:
processed_df2=processed_df[~(((processed_df['거래소코드']) ==1) & (processed_df['거래소코드'].value_counts()) == 1)]

mask = processed_df2['거래소코드'] == processed_df2['거래소코드'].shift(-1)
processed_df2['t-2감사의견코드'] = np.where(mask,processed_df2['t-1감사의견코드'].shift(-1), np.nan)

In [453]:
processed_df2['t-2감사의견코드'].value_counts()

t-2감사의견코드
0.0    146139
1.0      2298
Name: count, dtype: int64

In [454]:
processed_df2.dropna(subset='t-2감사의견코드',axis=0,inplace=True)

In [455]:
# processed_df.to_csv('t-2_raw_외감.csv',index=False,encoding='euc-kr')